File: app/src/main/java/com/zeapo/pwdstore/utils/PasswordRepository.java
package com.zeapo.pwdstore.utils;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.util.Log;

import org.apache.commons.io.filefilter.FileFilterUtils;
import org.eclipse.jgit.api.Git;
import org.eclipse.jgit.lib.Repository;
import org.eclipse.jgit.lib.StoredConfig;
import org.eclipse.jgit.storage.file.FileRepositoryBuilder;
import org.eclipse.jgit.transport.RefSpec;
import org.eclipse.jgit.transport.RemoteConfig;
import org.eclipse.jgit.transport.URIish;

import java.io.File;
import java.io.FileFilter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.Set;

import static java.util.Collections.sort;

public class PasswordRepository {

    private static Repository repository;

    protected PasswordRepository() {
    }

    /**
     * Returns the git repository
     *
     * @param localDir needed only on the creation
     * @return the git repository
     */
    public static Repository getRepository(File localDir) {
        if (repository == null && localDir != null) {
            FileRepositoryBuilder builder = new FileRepositoryBuilder();
            try {
                repository = builder.setGitDir(localDir)
                        .readEnvironment()
                        .build();
            } catch (Exception e) {
                e.printStackTrace();
                return null;
            }
        }
        return repository;
    }

    public static boolean isInitialized() {
        return repository != null;
    }

    public static void createRepository(File localDir) throws Exception {
        localDir.delete();

        Git.init().setDirectory(localDir).call();
        getRepository(localDir);
    }

    // TODO add multiple remotes support for pull/push
    public static void addRemote(String name, String url, Boolean replace) {
        StoredConfig storedConfig = repository.getConfig();
        Set<String> remotes = storedConfig.getSubsections("remote");

        if (!remotes.contains(name)) {
            try {
                URIish uri = new URIish(url);
                RefSpec refSpec = new RefSpec("+refs/head/*:refs/remotes/" + name + "/*");

                RemoteConfig remoteConfig = new RemoteConfig(storedConfig, name);
                remoteConfig.addFetchRefSpec(refSpec);
                remoteConfig.addPushRefSpec(refSpec);
                remoteConfig.addURI(uri);
                remoteConfig.addPushURI(uri);

                remoteConfig.update(storedConfig);

                storedConfig.save();
            } catch (Exception e) {
                e.printStackTrace();
            }
        } else if (replace) {
            try {
                URIish uri = new URIish(url);

                RemoteConfig remoteConfig = new RemoteConfig(storedConfig, name);
                // remove the first and eventually the only uri
                if (remoteConfig.getURIs().size() > 0) {
                    remoteConfig.removeURI(remoteConfig.getURIs().get(0));
                }
                if (remoteConfig.getPushURIs().size() > 0) {
                    remoteConfig.removePushURI(remoteConfig.getPushURIs().get(0));
                }

                remoteConfig.addURI(uri);
                remoteConfig.addPushURI(uri);

                remoteConfig.update(storedConfig);

                storedConfig.save();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public static void closeRepository() {
        if (repository != null) repository.close();
        repository = null;
    }

    public static File getRepositoryDirectory(Context context) {
        File dir = null;
        SharedPreferences settings = PreferenceManager.getDefaultSharedPreferences(context.getApplicationContext());

        if (settings.getBoolean("git_external", false)) {
            String external_repo = settings.getString("git_external_repo", null);
            if (external_repo != null) {
                dir = new File(external_repo);
            }
        } else {
            dir = new File(context.getFilesDir() + "/store");
        }

        return dir;
    }

    public static Repository initialize(Context context) {
        File dir = getRepositoryDirectory(context);
        SharedPreferences settings = PreferenceManager.getDefaultSharedPreferences(context.getApplicationContext());

        if (dir == null) {
            return null;
        }

        // uninitialize the repo if the dir does not exist or is absolutely empty
        if (!dir.exists() || !dir.isDirectory() || dir.listFiles().length == 0) {
            settings.edit().putBoolean("repository_initialized", false).apply();
        } else {
            settings.edit().putBoolean("repository_initialized", true).apply();
        }

        // create the repository static variable in PasswordRepository
        return PasswordRepository.getRepository(new File(dir.getAbsolutePath() + "/.git"));
    }

    /**
     * Gets the password items in the root directory
     *
     * @return a list of passwords in the root direcotyr
     */
    public static ArrayList<PasswordItem> getPasswords(File rootDir, PasswordSortOrder sortOrder) {
        return getPasswords(rootDir, rootDir, sortOrder);
    }

    /**
     * Gets the .gpg files in a directory
     *
     * @param path the directory path
     * @return the list of gpg files in that directory
     */
    public static ArrayList<File> getFilesList(File path) {
        if (path == null || !path.exists()) return new ArrayList<>();

        Log.d("REPO", "current path: " + path.getPath());
        List<File> directories = Arrays.asList(path.listFiles((FileFilter) FileFilterUtils.directoryFileFilter()));
        List<File> files = Arrays.asList(path.listFiles((FileFilter) FileFilterUtils.suffixFileFilter(".gpg")));

        ArrayList<File> items = new ArrayList<>();
        items.addAll(directories);
        items.addAll(files);

        return items;
    }

    /**
     * Gets the passwords (PasswordItem) in a directory
     *
     * @param path the directory path
     * @return a list of password items
     */
    public static ArrayList<PasswordItem> getPasswords(File path, File rootDir, PasswordSortOrder sortOrder) {
        //We need to recover the passwords then parse the files
        ArrayList<File> passList = getFilesList(path);

        if (passList.size() == 0) return new ArrayList<>();

        ArrayList<PasswordItem> passwordList = new ArrayList<>();

        for (File file : passList) {
            if (file.isFile()) {
                if (!file.isHidden()) {
                    passwordList.add(PasswordItem.newPassword(file.getName(), file, rootDir));
                }
            } else {
                if (!file.isHidden()) {
                    passwordList.add(PasswordItem.newCategory(file.getName(), file, rootDir));
                }
            }
        }
        sort(passwordList, sortOrder.comparator);
        return passwordList;
    }

    /**
     * Sets the git user name
     *
     * @param username username
     */
    public static void setUserName(String username) {
        setStringConfig("user", null, "name", username);
    }

    /**
     * Sets the git user email
     *
     * @param email email
     */
    public static void setUserEmail(String email) {
        setStringConfig("user", null, "email", email);
    }

    /**
     * Sets a git config value
     *
     * @param section    config section name
     * @param subsection config subsection name
     * @param name       config name
     * @param value      the value to be set
     */
    private static void setStringConfig(String section, String subsection, String name, String value) {
        if (isInitialized()) {
            StoredConfig config = repository.getConfig();
            config.setString(section, subsection, name, value);
            try {
                config.save();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public enum PasswordSortOrder {

        FOLDER_FIRST(new Comparator<PasswordItem>() {
            @Override
            public int compare(PasswordItem p1, PasswordItem p2) {
                return (p1.getType() + p1.getName())
                        .compareToIgnoreCase(p2.getType() + p2.getName());
            }
        }),

        INDEPENDENT(new Comparator<PasswordItem>() {
            @Override
            public int compare(PasswordItem p1, PasswordItem p2) {
                return p1.getName().compareToIgnoreCase(p2.getName());
            }
        }),

        FILE_FIRST(new Comparator<PasswordItem>() {
            @Override
            public int compare(PasswordItem p1, PasswordItem p2) {
                return (p2.getType() + p1.getName())
                        .compareToIgnoreCase(p1.getType() + p2.getName());
            }
        })

        ;

        private Comparator<PasswordItem> comparator;

        PasswordSortOrder(Comparator<PasswordItem> comparator) {
            this.comparator = comparator;
        }

        public static PasswordSortOrder getSortOrder(SharedPreferences settings) {
            return valueOf(settings.getString("sort_order", FOLDER_FIRST.name()));
        }
    }
}

File: app/src/main/java/com/zeapo/pwdstore/utils/PasswordRepository.kt
/*
 * Copyright © 2014-2020 The Android Password Store Authors. All Rights Reserved.
 * SPDX-License-Identifier: GPL-3.0-only
 */
package com.zeapo.pwdstore.utils

import android.os.Build
import androidx.annotation.RequiresApi
import androidx.core.content.edit
import com.github.michaelbull.result.getOrElse
import com.github.michaelbull.result.onFailure
import com.github.michaelbull.result.runCatching
import com.zeapo.pwdstore.Application
import java.io.File
import java.io.FileFilter
import java.nio.file.Files
import java.nio.file.LinkOption
import org.eclipse.jgit.api.Git
import org.eclipse.jgit.lib.Repository
import org.eclipse.jgit.storage.file.FileRepositoryBuilder
import org.eclipse.jgit.transport.RefSpec
import org.eclipse.jgit.transport.RemoteConfig
import org.eclipse.jgit.transport.URIish
import org.eclipse.jgit.util.FS
import org.eclipse.jgit.util.FS_POSIX_Java6

object PasswordRepository {

    @RequiresApi(Build.VERSION_CODES.O)
    private class FS_POSIX_Java6_with_optional_symlinks : FS_POSIX_Java6() {

        override fun supportsSymlinks() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.O

        override fun isSymLink(file: File) = Files.isSymbolicLink(file.toPath())

        override fun readSymLink(file: File) = Files.readSymbolicLink(file.toPath()).toString()

        override fun createSymLink(source: File, target: String) {
            val sourcePath = source.toPath()
            if (Files.exists(sourcePath, LinkOption.NOFOLLOW_LINKS))
                Files.delete(sourcePath)
            Files.createSymbolicLink(sourcePath, File(target).toPath())
        }
    }

    @RequiresApi(Build.VERSION_CODES.O)
    private class Java7FSFactory : FS.FSFactory() {

        override fun detect(cygwinUsed: Boolean?): FS {
            return FS_POSIX_Java6_with_optional_symlinks()
        }
    }

    private var repository: Repository? = null
    private val settings by lazy(LazyThreadSafetyMode.NONE) { Application.instance.sharedPrefs }
    private val filesDir
        get() = Application.instance.filesDir

    /**
     * Returns the git repository
     *
     * @param localDir needed only on the creation
     * @return the git repository
     */
    @JvmStatic
    fun getRepository(localDir: File?): Repository? {
        if (repository == null && localDir != null) {
            val builder = FileRepositoryBuilder()
            repository = runCatching {
                builder.run {
                    gitDir = localDir
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                        fs = Java7FSFactory().detect(null)
                    }
                    readEnvironment()
                }.build()
            }.getOrElse { e ->
                e.printStackTrace()
                null
            }
        }
        return repository
    }

    @JvmStatic
    val isInitialized: Boolean
        get() = repository != null

    @JvmStatic
    fun isGitRepo(): Boolean {
        if (repository != null) {
            return repository!!.objectDatabase.exists()
        }
        return false
    }

    @JvmStatic
    @Throws(Exception::class)
    fun createRepository(localDir: File) {
        localDir.delete()

        Git.init().setDirectory(localDir).call()
        getRepository(localDir)
    }

    // TODO add multiple remotes support for pull/push
    @JvmStatic
    fun addRemote(name: String, url: String, replace: Boolean = false) {
        val storedConfig = repository!!.config
        val remotes = storedConfig.getSubsections("remote")

        if (!remotes.contains(name)) {
            runCatching {
                val uri = URIish(url)
                val refSpec = RefSpec("+refs/head/*:refs/remotes/$name/*")

                val remoteConfig = RemoteConfig(storedConfig, name)
                remoteConfig.addFetchRefSpec(refSpec)
                remoteConfig.addPushRefSpec(refSpec)
                remoteConfig.addURI(uri)
                remoteConfig.addPushURI(uri)

                remoteConfig.update(storedConfig)

                storedConfig.save()
            }.onFailure { e ->
                e.printStackTrace()
            }
        } else if (replace) {
            runCatching {
                val uri = URIish(url)

                val remoteConfig = RemoteConfig(storedConfig, name)
                // remove the first and eventually the only uri
                if (remoteConfig.urIs.size > 0) {
                    remoteConfig.removeURI(remoteConfig.urIs[0])
                }
                if (remoteConfig.pushURIs.size > 0) {
                    remoteConfig.removePushURI(remoteConfig.pushURIs[0])
                }

                remoteConfig.addURI(uri)
                remoteConfig.addPushURI(uri)

                remoteConfig.update(storedConfig)

                storedConfig.save()
            }.onFailure { e ->
                e.printStackTrace()
            }
        }
    }

    @JvmStatic
    fun closeRepository() {
        if (repository != null) repository!!.close()
        repository = null
    }

    @JvmStatic
    fun getRepositoryDirectory(): File {
        return if (settings.getBoolean(PreferenceKeys.GIT_EXTERNAL, false)) {
            val externalRepo = settings.getString(PreferenceKeys.GIT_EXTERNAL_REPO)
            if (externalRepo != null)
                File(externalRepo)
            else
                File(filesDir.toString(), "/store")
        } else {
            File(filesDir.toString(), "/store")
        }
    }

    @JvmStatic
    fun initialize(): Repository? {
        val dir = getRepositoryDirectory()
        // uninitialize the repo if the dir does not exist or is absolutely empty
        settings.edit {
            if (!dir.exists() || !dir.isDirectory || requireNotNull(dir.listFiles()).isEmpty()) {
                putBoolean(PreferenceKeys.REPOSITORY_INITIALIZED, false)
            } else {
                putBoolean(PreferenceKeys.REPOSITORY_INITIALIZED, true)
            }
        }

        // create the repository static variable in PasswordRepository
        return getRepository(File(dir.absolutePath + "/.git"))
    }

    /**
     * Gets the .gpg files in a directory
     *
     * @param path the directory path
     * @return the list of gpg files in that directory
     */
    @JvmStatic
    fun getFilesList(path: File?): ArrayList<File> {
        if (path == null || !path.exists()) return ArrayList()

        val directories = (path.listFiles(FileFilter { pathname -> pathname.isDirectory })
            ?: emptyArray()).toList()
        val files = (path.listFiles(FileFilter { pathname -> pathname.extension == "gpg" })
            ?: emptyArray()).toList()

        val items = ArrayList<File>()
        items.addAll(directories)
        items.addAll(files)

        return items
    }

    /**
     * Gets the passwords (PasswordItem) in a directory
     *
     * @param path the directory path
     * @return a list of password items
     */
    @JvmStatic
    fun getPasswords(path: File, rootDir: File, sortOrder: PasswordSortOrder): ArrayList<PasswordItem> {
        // We need to recover the passwords then parse the files
        val passList = getFilesList(path).also { it.sortBy { f -> f.name } }
        val passwordList = ArrayList<PasswordItem>()
        val showHidden = settings.getBoolean(PreferenceKeys.SHOW_HIDDEN_CONTENTS, false)

        if (passList.size == 0) return passwordList
        if (!showHidden) {
            passList.filter { !it.isHidden }.toCollection(passList.apply { clear() })
        }
        passList.forEach { file ->
            passwordList.add(if (file.isFile) {
                PasswordItem.newPassword(file.name, file, rootDir)
            } else {
                PasswordItem.newCategory(file.name, file, rootDir)
            })
        }
        passwordList.sortWith(sortOrder.comparator)
        return passwordList
    }
}
